File: examples/nostr-basics/03-subscribe-filter/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/gorilla/websocket"
	nostr "github.com/nbd-wtf/go-nostr"
)

func main() {
	relay := "ws://127.0.0.1:3100"
	if len(os.Args) > 1 {
		relay = os.Args[1]
	}
	const subscriptionID = "learn-nip01"

	filter := map[string]any{"kinds": []int{1}, "limit": 5}
	if author := os.Getenv("NOSTR_AUTHOR"); author != "" {
		filter["authors"] = []string{author}
	}

	conn, _, err := websocket.DefaultDialer.Dial(relay, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()
	if err := conn.WriteJSON([]any{"REQ", subscriptionID, filter}); err != nil {
		panic(err)
	}
	fmt.Printf("sent REQ %s: %v\n", subscriptionID, filter)

	_ = conn.SetReadDeadline(time.Now().Add(15 * time.Second))
	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			if netError, ok := err.(interface{ Timeout() bool }); ok && netError.Timeout() {
				fmt.Println("listen window elapsed")
				break
			}
			panic(err)
		}

		var envelope []json.RawMessage
		if err := json.Unmarshal(data, &envelope); err != nil || len(envelope) == 0 {
			continue
		}
		var label string
		_ = json.Unmarshal(envelope[0], &label)
		switch label {
		case "EVENT":
			if len(envelope) < 3 {
				continue
			}
			var subID string
			_ = json.Unmarshal(envelope[1], &subID)
			if subID != subscriptionID {
				continue
			}
			var event nostr.Event
			if err := json.Unmarshal(envelope[2], &event); err != nil {
				panic(err)
			}
			fmt.Printf("EVENT %s kind=%d author=%s content=%q\n", event.ID, event.Kind, event.PubKey, event.Content)
		case "EOSE":
			fmt.Println("EOSE: historical events are complete")
		case "CLOSED":
			fmt.Println("CLOSED:", string(data))
			return
		case "NOTICE":
			fmt.Println("NOTICE:", string(data))
		}
	}

	_ = conn.WriteJSON([]any{"CLOSE", subscriptionID})
}

File: examples/nostr-basics/03-subscribe-filter/rust/src/main.rs
use std::error::Error;
use std::time::Duration;

use futures_util::{SinkExt, StreamExt};
use nostr::Event;
use serde_json::{json, Value};
use tokio::time::timeout;
use tokio_tungstenite::{connect_async, tungstenite::Message};

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let relay = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "ws://127.0.0.1:3100".to_string());
    let subscription_id = "learn-nip01";
    let mut filter = json!({"kinds": [1], "limit": 5});
    if let Ok(author) = std::env::var("NOSTR_AUTHOR") {
        filter["authors"] = json!([author]);
    }

    let (mut ws, _) = connect_async(&relay).await?;
    ws.send(Message::Text(
        json!(["REQ", subscription_id, filter]).to_string().into(),
    ))
    .await?;
    println!("sent REQ {subscription_id}: {filter}");

    let listening = timeout(Duration::from_secs(15), async {
        while let Some(message) = ws.next().await {
            let Message::Text(text) = message? else {
                continue;
            };
            let value: Value = serde_json::from_str(text.as_ref())?;
            let Some(array) = value.as_array() else {
                continue;
            };
            match array.first().and_then(Value::as_str) {
                Some("EVENT") if array.get(1).and_then(Value::as_str) == Some(subscription_id) => {
                    if let Some(raw_event) = array.get(2) {
                        let event: Event = serde_json::from_value(raw_event.clone())?;
                        println!(
                            "EVENT {} kind={} author={} content={:?}",
                            event.id.to_hex(),
                            event.kind.as_u16(),
                            event.pubkey.to_hex(),
                            event.content
                        );
                    }
                }
                Some("EOSE") => println!("EOSE: historical events are complete"),
                Some("CLOSED") => {
                    println!("CLOSED: {value}");
                    break;
                }
                Some("NOTICE") => println!("NOTICE: {value}"),
                _ => {}
            }
        }
        Ok::<(), Box<dyn Error>>(())
    })
    .await;

    if listening.is_err() {
        println!("listen window elapsed; sending CLOSE");
    }
    ws.send(Message::Text(
        json!(["CLOSE", subscription_id]).to_string().into(),
    ))
    .await?;
    Ok(())
}
